File: ruby/rpc_server.rb
#!/usr/bin/env ruby
# encoding: utf-8

require "bunny"

conn = Bunny.new
conn.start

ch   = conn.create_channel

class FibonacciServer

  def initialize(ch)
    @ch = ch
  end

  def start(queue_name)
    @q = @ch.queue(queue_name)
    @x = @ch.default_exchange

    @q.subscribe(:block => true) do |delivery_info, properties, payload|
      n = payload.to_i
      r = self.class.fib(n)

      puts " [.] fib(#{n})"

      @x.publish(r.to_s, :routing_key => properties.reply_to, :correlation_id => properties.correlation_id)
    end
  end


  def self.fib(n)
    case n
    when 0 then 0
    when 1 then 1
    else
      fib(n - 1) + fib(n - 2)
    end
  end
end

begin
  server = FibonacciServer.new(ch)
  " [x] Awaiting RPC requests"
  server.start("rpc_queue")
rescue Interrupt => _
  ch.close
  conn.close

  exit(0)
end

File: perl/rpc_server.pl
#!/usr/bin/perl

use strict;
use warnings;

$|++;
use AnyEvent;
use Net::RabbitFoot;

my $conn = Net::RabbitFoot->new()->load_xml_spec()->connect(
    host => 'localhost',
    port => 5672,
    user => 'guest',
    pass => 'guest',
    vhost => '/',
);

my $channel = $conn->open_channel();

$channel->declare_queue(queue => 'rpc_queue');

sub fib {
    my $n = shift;
    my ($a, $b) = (0, 1);
    for (1..$n) {
        ($a, $b) = ($b, $a + $b);
    }
    return $a;
}

sub on_request {
    my $var = shift;
    my $body = $var->{body}->{payload};
    my $props = $var->{header};

    my $n = $body;
    print " [.] fib($n)\n";
    my $response = fib($n);

    $channel->publish(
        exchange => '',
        routing_key => $props->{reply_to},
        header => {
            correlation_id => $props->{correlation_id},
        },
        body => $response,
    );

    $channel->ack();
}

$channel->qos(prefetch_count => 1);
$channel->consume(
    on_consume => \&on_request,
);

print " [x] Awaiting RPC requests\n";

# Wait forever
AnyEvent->condvar->recv;
